pkg/cli/tty_unix.go

```diff
@@ -0,0 +1,21 @@
+// +build !windows,!plan9,!js
+
+package cli
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+)
+
+func (t *aTTY) NotifySignals() <-chan os.Signal {
+	// This implicitly catches every signal regardless of whether it is
+	// currently ignored.
+	t.sigCh = make(chan os.Signal, sigsChanBufferSize)
+	signal.Notify(t.sigCh)
+	// TODO: Remove this if, and when, job control is implemented. This must
+	// be done after calling signal.Notify() or it won't have any effect. See
+	// https://github.com/elves/elvish/issues/988.
+	signal.Ignore(syscall.SIGTTIN, syscall.SIGTTOU, syscall.SIGTSTP)
```

Comment:
This seems to be the wrong place to ignore the signal. The editor is only active in interactive mode, but the signal ignoring behavior should also apply to the script mode. This should be done somewhere in `pkg/shell`.

Ideally, we should do this between `fork(3)` and `exec(3)`, but Go only offers a combined [`os.ForkExec`](https://godoc.org/syscall#ForkExec) API (for good reasons; threads and forks require a lot of care to interoperate), and there is a limited amount of things you can configure with [ProcAttr](https://godoc.org/syscall#ProcAttr).

Elvish should maintain its own version of `ForkExec` when there is a more compelling use case and someone has all the time and caution in the world, but for this particular case a blanket `signal.Ignore` seems sufficient and harmless.

Reply:
This is the right place as it only applies to interactive mode. If you're running a script non-interactively (e.g., via cron) then it won't have a controlling tty and SIGTSTP isn't an issue. If you run an elvish script as an external command from an interactive elvish session then this change means it will (or should) be run with SIGTSTP ignored and will propagate that to any processes it runs. Or it would except I see that elvish resets SIGTSTP to its default behavior at startup -- which it shouldn't. I'll fix that and push an updated version of this PR.

Reply:
Oh, you are right.

It's still much cleaner to do that in `pkg/shell` in the interactive branch, both because that's simpler, and because this method, `NotifySignals`, is not really supposed to have the side effect of ignoring signals when the interactive editor is not active.

Reply:
We actually have to ignore the job-control signals in both locations. That's because the Go `signal.Notify()` API is suboptimal. If we don't pass an explicit list of signals it ignores that a given signal is being ignored and always enables the signal. The only way to avoid the need to call `signal.Ignore()` in this function is by explicitly specifying the set of signals for which we want notifications. That set must therefore exclude the three tty job-control signals. We therefore have to write `signal.Notify(syscall, SIGABRT, syscall.SIGALRM, etc.)`.